File: releasing/commands/src/tarball/summarizeTarball.ts
import { createHash } from 'node:crypto'
import { gunzipSync } from 'node:zlib'

import { PnpmError } from '@pnpm/error'
import tar from 'tar-stream'

import { extractBundledDependencies, type PublishSummary } from './publishSummary.js'
import { createTarballFilename } from './safeTarballFilename.js'

interface TarballManifest {
  _id?: string
  name?: string
  version?: string
  bundledDependencies?: unknown
  bundleDependencies?: unknown
  dependencies?: Record<string, unknown>
}

/**
 * Parse a packed (gzipped or plain) tarball buffer and return the same
 * {@link PublishSummary} shape that `pnpm publish --json` emits.
 *
 * Used when we hold the tarball bytes already and need a summary without
 * re-packing — e.g. inspecting a staged publish via `pnpm stage download`.
 *
 * @throws {@link PnpmError} with code `STAGE_TARBALL_MANIFEST_NOT_FOUND` when the tarball
 *   does not contain `package/package.json`, or when that file is unparseable JSON.
 */
export async function summarizeTarball (tarballData: Buffer): Promise<PublishSummary> {
  const extract = tar.extract()
  const files: Array<{ path: string }> = []
  const bundled = new Set<string>()
  let manifest: TarballManifest | undefined
  let entryCount = 0
  let unpackedSize = 0

  await new Promise<void>((resolve, reject) => {
    extract.on('entry', (header, stream, next) => {
      const chunks: Buffer[] = []
      if (header.type === 'file') {
        entryCount++
        unpackedSize += header.size ?? 0
        files.push({ path: header.name.replace(/^package\//, '') })
        const bundledMatch = /^package\/node_modules\/((?:@[^/]+\/)?[^/]+)/.exec(header.name)
        if (bundledMatch?.[1]) {
          bundled.add(bundledMatch[1])
        }
      }
      if (header.name === 'package/package.json') {
        stream.on('data', (chunk: Buffer) => chunks.push(Buffer.from(chunk)))
      }
      stream.on('error', reject)
      stream.on('end', () => {
        if (header.name === 'package/package.json') {
          try {
            manifest = JSON.parse(Buffer.concat(chunks).toString())
          } catch (error: unknown) {
            reject(error)
            return
          }
        }
        next()
      })
      stream.resume()
    })
    extract.on('error', reject)
    extract.on('finish', resolve)
    extract.end(maybeGunzip(tarballData))
  })

  if (!manifest?.name || !manifest.version) {
    throw new PnpmError('STAGE_TARBALL_MANIFEST_NOT_FOUND', 'Could not read package.json from tarball')
  }

  files.sort((a, b) => a.path.localeCompare(b.path, 'en'))
  return {
    id: manifest._id ?? `${manifest.name}@${manifest.version}`,
    name: manifest.name,
    version: manifest.version,
    size: tarballData.byteLength,
    unpackedSize,
    shasum: createHash('sha1').update(tarballData).digest('hex'),
    integrity: `sha512-${createHash('sha512').update(tarballData).digest('base64')}`,
    filename: createTarballFilename({ name: manifest.name, version: manifest.version }),
    files,
    entryCount,
    bundled: bundled.size > 0 ? Array.from(bundled).sort() : extractBundledDependencies(manifest),
  }
}

function maybeGunzip (tarballData: Buffer): Buffer {
  try {
    return gunzipSync(tarballData)
  } catch {
    return tarballData
  }
}

File: pnpm/crates/cli/src/cli_args/stage/summarize_tarball.rs
//! Summarize a packed tarball's bytes into the [`PublishSummary`] shape that
//! `pnpm publish --json` emits — used by `pnpm stage download` to describe a
//! staged tarball without re-packing it.

use std::{collections::BTreeSet, io::Read};

use flate2::read::GzDecoder;
use miette::{Context, IntoDiagnostic};
use pacquet_pack::sort_paths_en_locale;
use pacquet_publish::{PackedPkgInfo, PublishSummary, create_publish_summary};
use pacquet_resolving_parse_wanted_dependency::is_valid_old_npm_package_name;
use serde_json::Value;

use super::StageError;

/// Parse a packed (gzipped or plain) tarball and return its
/// [`PublishSummary`]. The tarball must contain a parseable
/// `package/package.json` with a name and version.
pub(super) fn summarize_tarball(tarball_data: &[u8]) -> miette::Result<PublishSummary> {
    let tar_bytes = maybe_gunzip(tarball_data)?;
    let mut archive = tar::Archive::new(tar_bytes.as_slice());
    let mut files: Vec<String> = Vec::new();
    let mut bundled: BTreeSet<String> = BTreeSet::new();
    let mut manifest: Option<Value> = None;
    let mut unpacked_size: u64 = 0;

    let entries =
        archive.entries().into_diagnostic().wrap_err("read the staged tarball's entries")?;
    for entry in entries {
        let mut entry = entry.into_diagnostic().wrap_err("read a staged tarball entry")?;
        let path = String::from_utf8_lossy(&entry.path_bytes()).into_owned();
        if entry.header().entry_type().is_file() {
            unpacked_size += entry.header().size().unwrap_or(0);
            files.push(path.strip_prefix("package/").unwrap_or(&path).to_owned());
            if let Some(name) = bundled_dependency_name(&path) {
                bundled.insert(name);
            }
        }
        if path == "package/package.json" {
            let mut text = String::new();
            entry
                .read_to_string(&mut text)
                .into_diagnostic()
                .wrap_err("read package/package.json from the staged tarball")?;
            manifest =
                Some(serde_json::from_str(&text).map_err(|_| StageError::TarballManifestNotFound)?);
        }
    }

    let manifest = manifest.ok_or(StageError::TarballManifestNotFound)?;
    let name = manifest_string(&manifest, "name");
    let version = manifest_string(&manifest, "version");
    if name.is_empty() || version.is_empty() {
        return Err(StageError::TarballManifestNotFound.into());
    }

    sort_paths_en_locale(&mut files);
    let filename = create_tarball_filename(&name, &version, None)?;
    let mut summary = create_publish_summary(
        &PackedPkgInfo {
            published_manifest: &manifest,
            tarball_path: &filename,
            contents: &files,
            unpacked_size,
        },
        tarball_data,
    );
    // A registry-packed tarball's manifest carries an `_id`; prefer it over
    // the derived `name@version`, matching pnpm's `summarizeTarball`.
    if let Some(id) = manifest.get("_id").and_then(Value::as_str) {
        id.clone_into(&mut summary.id);
    }
    if !bundled.is_empty() {
        summary.bundled = bundled.into_iter().collect();
    }
    Ok(summary)
}

/// The safe tarball basename `<normalized-name>-<version>[-<suffix>].tgz`,
/// after validating that the name and version cannot smuggle path segments.
pub(super) fn create_tarball_filename(
    name: &str,
    version: &str,
    suffix: Option<&str>,
) -> Result<String, StageError> {
    if !is_valid_old_npm_package_name(name) {
        return Err(StageError::InvalidPackageName { name: name.to_owned() });
    }
    if version.parse::<node_semver::Version>().is_err() {
        return Err(StageError::InvalidPackageVersion { version: version.to_owned() });
    }
    let suffix = suffix.map(|suffix| format!("-{suffix}")).unwrap_or_default();
    let filename = format!("{}-{version}{suffix}.tgz", normalize_package_name(name));
    // The name/version validation above should already exclude separators;
    // reject outright if a validated component still smuggled one in.
    if filename.contains(['/', '\\', ':']) {
        return Err(StageError::InvalidTarballFilename { filename });
    }
    Ok(filename)
}

/// `@scope/name` → `scope-name`: drop the first `@`, turn the first `/` into
/// a `-`, exactly like pnpm's `normalizePackageName`.
fn normalize_package_name(name: &str) -> String {
    name.replacen('@', "", 1).replacen('/', "-", 1)
}

/// The bundled-dependency name of a `package/node_modules/...` tarball entry:
/// the first path segment (or two for a scoped package) after
/// `node_modules/`.
fn bundled_dependency_name(path: &str) -> Option<String> {
    let rest = path.strip_prefix("package/node_modules/")?;
    let mut segments = rest.split('/');
    let first = segments.next().filter(|segment| !segment.is_empty())?;
    if let Some(scope) = first.strip_prefix('@') {
        if scope.is_empty() {
            return None;
        }
        let second = segments.next().filter(|segment| !segment.is_empty())?;
        return Some(format!("{first}/{second}"));
    }
    Some(first.to_owned())
}

/// Cap on the decompressed tarball; the bytes come from the registry, so a
/// gzip bomb must not inflate past this into memory.
const MAX_DECOMPRESSED_TARBALL_BYTES: u64 = 512 * 1024 * 1024;

/// Gunzip `data`, falling back to the raw bytes when it is not gzipped.
/// Decompression past [`MAX_DECOMPRESSED_TARBALL_BYTES`] is an error.
fn maybe_gunzip(data: &[u8]) -> Result<Vec<u8>, StageError> {
    let mut decoder = GzDecoder::new(data).take(MAX_DECOMPRESSED_TARBALL_BYTES + 1);
    let mut decompressed = Vec::new();
    match decoder.read_to_end(&mut decompressed) {
        Ok(_) if decompressed.len() as u64 > MAX_DECOMPRESSED_TARBALL_BYTES => {
            Err(StageError::RequestFailed {
                operation: "read the staged tarball".to_owned(),
                reason: format!(
                    "tarball exceeded {MAX_DECOMPRESSED_TARBALL_BYTES} bytes when decompressed",
                ),
            })
        }
        Ok(_) => Ok(decompressed),
        Err(_) => Ok(data.to_vec()),
    }
}

fn manifest_string(manifest: &Value, key: &str) -> String {
    manifest.get(key).and_then(Value::as_str).unwrap_or_default().to_owned()
}
